### PR TITLE
[PM-4530] Fix sso in snap desktop

### PR DESCRIPTION
**Pull request by quexten**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-4530

Server PR: https://github.com/bitwarden/server/pull/4649

## 📔 Objective

Fix sso in snap desktop [and appimage, electron-dev[windows/mac/linux]] by porting over the localhost listener behaviour from the cli.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/desktop/src/platform/services/sso-localhost-callback.service.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        const urlString = "http://localhost" + req.url;
+        const url = new URL(urlString);
+        const code = url.searchParams.get("code");
+        const receivedState = url.searchParams.get("state");
+        const orgIdentifier = this.getOrgIdentifierFromState(receivedState);
+        res.setHeader("Content-Type", "text/html");
+        if (code != null && receivedState != null && this.checkState(receivedState, state)) {
+          res.writeHead(200);
+          res.end(
+            "<html><head><title>Success | Bitwarden CLI</title></head><body>" +
```
Comment (dani-garcia) [inline apps/desktop/src/platform/services/sso-localhost-callback.service.ts:55]: These should reference the desktop app and not the CLI. Same for the 400 error case

Comment (quexten) [reply apps/desktop/src/platform/services/sso-localhost-callback.service.ts]: Fixed

**Review by dani-garcia — APPROVED** [review]

File: apps/desktop/src/platform/services/sso-localhost-callback.service.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            });
+          });
+          foundPort = true;
+          break;
+        } catch {
+          // Ignore error since we run the same command up to 5 times.
+        }
+      }
+      if (!foundPort) {
+        reject();
```
Comment (jlf0dev) [inline apps/desktop/src/platform/services/sso-localhost-callback.service.ts:103]: Whats the error handling situation here? Do we have a way of notifying the user we couldn't find a free port?

**Review by jlf0dev — COMMENTED** [review]
Just a small question. Overall looks great!

In the future it might be nice to pull the `bitwarden://sso-callback` into its own service as well and use DI to chose between the two, but definitely enhancement territory and out of scope here.

File: apps/desktop/src/platform/services/sso-localhost-callback.service.ts (hunk 2 above)
Comment (quexten) [reply apps/desktop/src/platform/services/sso-localhost-callback.service.ts:103]: Added an error message for this.

**Review by jlf0dev — APPROVED** [review]

**Review by dani-garcia — APPROVED** [review]

**Review by ike-kottlowski — APPROVED** [review]

**Pull request merged**